server/log_rpc_server.go

```diff
@@ -98,6 +98,14 @@ func (t *TrillianLogRPCServer) QueueLeaves(ctx context.Context, req *trillian.Qu
 	if err != nil {
 		return nil, err
 	}
+
+	// If we're draining the queue reject attempts to add new leaves. This is
+	// not the ideal place to enforce this restriction but it's currently
+	// the only really practical one.
```

Comment:
Why not enforce it in the same place as `TreeState_FROZEN`?
 - server/interceptor/interceptor.go marks a `QueueLeaves` request as `readonly=false`
 - change `trees.GetTree` to then check for `DRAINING` as well as `FROZEN` -- either is effectively readonly from a gRPC API perspective.

Reply:
I tried this first and it doesn't work because the storage API only has a readonly flag and you don't have an options struct for GetTree. Fixing this up would be an intrusive change to public APIs.

Reply:
How about this then:
 - add an extra `Internal bool // Tree will only be used for internal (non-API-driven) purposes` field to `trees.GetOpts`
 - set `Internal: true` in `SequencerManager.ExecutePass` (only)
 - in `trees.GetTree` allow `DRAINING` trees to be returned if `!opts.Readonly || opts.Internal`
Would that work?

Reply:
I don't think so. GetTrees is called 3 times - by the interceptor, the code in the RPC server and via storage. The last of these is the problem because you've lost the context of what is being done. You only know if it's a read or write transaction.  This wasn't previously a problem because there was nothing selective about write permission as FROZEN denies everything.

Reply:
OK, so the storage/*/*_storage.go invocations of `trees.GetTree` would all have to be `Internal: true` as well, and we'd be relying on the interceptor and RPC server policing the external boundaries correctly.

Reply:
I think that's more or less it. GetOpts doesn't make it into storage. It makes a new one and stuffs the readonly flag in it so it can call back to GetTrees. There's no easy way to set internal for one operation only. We should probably have had an options struct on the storage code but it's a breaking change to add it now.

Reply:
This has now been completely redone after the API changes were merged. All the code that handles state / access type validation is now in one place in trees.go and is much cleaner.